File: 프로그래머스/2/131130. 혼자 놀기의 달인/혼자 놀기의 달인.cpp
#include <string>
#include <vector>
#include <iostream>
#include <algorithm>

using namespace std;

bool compare(int a, int b) {
    return a > b;
}

int targeting(vector<int>& cards, vector<int>& visited, int i) {
    int count = 0;
    while(visited[i] != 1) {
        visited[i] = 1;
        count++;
        i = cards[i];
    }
    return count;
}

int solution(vector<int> cards) {
    cards.insert(cards.begin(), 0);
    int n = cards.size();
    vector<int> visited(n, 0);
    vector<int> counts;
    
    for(int i=1; i<n; i++) {
        if(visited[i] != 1) {
            counts.push_back(targeting(cards, visited, i));
        }
    }
    
    if(counts.size() < 2) {
        return 0;
    }
    else {
        sort(counts.begin(), counts.end(), compare);
        return counts[0] * counts[1];
    }
}

File: 프로그래머스/2/131130. 혼자 놀기의 달인/혼자 놀기의 달인.py
def solution(cards):
    cards = [-1] + cards
    v = [-1 for _ in range(len(cards))]
    groups = []

    def dfs(current):
        if v[current] != -1:
            groups.append(g)
            return
        
        g.append(cards[current])
        next = cards[current]
        v[current] = 1
        dfs(next)

    for i in range(1, len(cards)):
        if v[i] == -1:
            g = []
            dfs(i)
    
    groups.sort(key=lambda x:len(x))

    if len(groups) == 1: return 0
    return len(groups[-1]) * len(groups[-2])